examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -0,0 +1,434 @@
+            val = dicts[key][param + '_freq']
+            if not isinstance(val, (int, float)):
+                raise ValueError(
+                    "Frequencies must be numbers, got %s" % type(val))
+        # check freq bands
+        if (filt_params_noise['l_freq'] > filt_params_signal['l_freq'] or
+                filt_params_signal['h_freq'] > filt_params_noise['h_freq']):
+            raise ValueError('Wrongly specified frequency bands!\n'
+                    'The signal band-pass must be within the noise band-pass!')
+
+        self.freqs_signal = (filt_params_signal['l_freq'],
+                             filt_params_signal['h_freq'])
+        self.freqs_noise = (filt_params_noise['l_freq'],
+                            filt_params_noise['h_freq'])
+        self.filt_params_signal = filt_params_signal
+        self.filt_params_noise = filt_params_noise
+        self.sort_by_spectral_ratio = sort_by_spectral_ratio
+        if n_fft is None:
+            self.n_fft = int(sampling_freq)
+        else:
+            self.n_fft = int(n_fft)
+        self.picks_ = (Ellipsis if picks is None else picks) 
+        self.return_filtered = return_filtered
+        self.estimator = estimator
+        self.n_components = n_components
+        self.rank = rank
+        self.sampling_freq = sampling_freq
+        self.cov_method_params = cov_method_params
+
+    def _check_X(self, X):
+        """Check input data."""
+        if not isinstance(X, np.ndarray):
+            raise ValueError("X should be of type ndarray (got %s)."
+                             % type(X))
+        if X.ndim < 2:
+            raise ValueError('X must have at least 2 dimensions.')
+        elif X.ndim > 3:
+            raise ValueError('X must have at most 3 dimensions.')
+
+    def fit(self, X, y=None):
+        """Estimate the SSD decomposition on raw or epoched data.
+
+        Parameters
+        ----------
+        X : array, shape (n_channels, n_times) | shape (n_epochs,
+                n_channels, n_times)
+            The input data from which to estimate the SSD. Either 2D array
+            obtained from continous data or 3D array obtained from epoched
+            data.
+        y : None | array, shape (n_samples,)
+                    Used for scikit-learn compatibility.
+        Returns
+        -------
+        self : instance of SSD
+            Returns the modified instance.
+        """
+        self._check_X(X)
+        if X.ndim == 2:
+            X_aux = X[self.picks_, :].copy()
+
+        elif X.ndim == 3:
+            X_aux = X[:, self.picks_, :].copy()
+        else:
+            raise NotImplementedError()
+
+        X_signal = filter_data(
+            X_aux, self.sampling_freq, **self.filt_params_signal)
+        X_noise = filter_data(
+            X_aux, self.sampling_freq, **self.filt_params_noise)
+
+        X_noise -= X_signal 
+        if X.ndim == 3:
+            X_signal = np.hstack(X_signal)
+            X_noise = np.hstack(X_noise)
```

Comment:
@vpeterson this is something I have not decided about. I feel keeping it 3d before filtering aligns the workflow with what ```mne.Epochs.filter``` would be doing. On the other hand, concatenating all signals before can lead different results, depending on how data is preprocessed, if edge artifacts occur, etc. Any thoughts @larsoner @agramfort ?

Reply:
yes you should keep the data 3D for filtering otherwise you'll have bad edge artifacts